src/scp/QuorumIntersectionChecker.cpp

```diff
@@ -0,0 +1,1106 @@
+// elsewhere in stellar-core so there's a little work up front converting
+// representations.
+//
+// Remaining details of the implementation are noted as we go, but the above
+// explanation ought to give you a good idea what you're looking at.
+
+namespace stellar
+{
+
+struct QBitSet;
+using QGraph = std::vector<QBitSet>;
+
+// A QBitSet is the "fast" representation of a SCPQuorumSet. It includes both a
+// BitSet of its own nodes and a summary BitSet of the union of its innerSets.
+struct QBitSet {
+    uint32_t mThreshold;
+    BitSet mNodes;
+    QGraph mInnerSets;
+
+    // Union of mNodes and all inner elements of mInnerSets.
+    BitSet mAllSuccessors;
+
+    QBitSet()
+    {}
+    QBitSet(uint32_t threshold,
+            BitSet const& nodes,
+            QGraph const& innerSets)
+        : mThreshold(threshold),
+          mNodes(nodes),
+          mInnerSets(innerSets)
+    {
+        getSuccessors(mAllSuccessors);
+    }
+
+    void log(size_t indent=0) const {
+        std::string s(indent, ' ');
+        CLOG(DEBUG, "SCP") << s
+                           << "QBitSet: thresh=" << mThreshold
+                           << "/" << (mNodes.count() + mInnerSets.size())
+                           << " validators=" << mNodes;
+        for (auto const& inner : mInnerSets) {
+            inner.log(indent+4);
+        }
+    }
+
+    void getSuccessors(BitSet &out) const {
+        out |= mNodes;
+        for (auto const& i : mInnerSets) {
+            i.getSuccessors(out);
+        }
+    }
+};
+
+// This is a completely stock implementation of Tarjan's algorithm for
+// calculating strongly connected components. Like "read off of wikipedia"
+// stock. Go have a look!
+//
+// https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm
+struct TarjanSCCCalculator {
+
+    struct SCCNode {
+        ssize_t mIndex = {-1};
+        ssize_t mLowLink = {-1};
+        bool mOnStack = {false};
+    };
+
+    std::vector<SCCNode> mNodes;
+    std::vector<size_t> mStack;
+    size_t mIndex = {0};
+    std::vector<BitSet> mSCCs;
+    QGraph const& mGraph;
+
+    TarjanSCCCalculator(QGraph const& graph)
+        : mGraph(graph)
+    {
+    }
+
+    void calculateSCCs() {
+        mNodes.clear();
+        mStack.clear();
+        mIndex = 0;
+        mSCCs.clear();
+        for (size_t i = 0; i < mGraph.size(); ++i)
+        {
+            mNodes.emplace_back(SCCNode{});
+        }
+        for (size_t i = 0; i < mGraph.size(); ++i)
+        {
+            if (mNodes.at(i).mIndex == -1)
+            {
+                scc(i);
+            }
+        }
+    }
+
+    void scc(size_t i) {
+        auto & v = mNodes.at(i);
+        v.mIndex = mIndex;
+        v.mLowLink = mIndex;
+        mIndex++;
+        mStack.push_back(i);
+        v.mOnStack = true;
+
+        BitSet const& succ = mGraph.at(i).mAllSuccessors;
+        for (size_t j = 0; succ.next_set(j); ++j) {
+            CLOG(DEBUG, "SCP") << "edge: " << i << " -> " << j;
```

Comment:
is j here always 0 on first iteration?

Reply:
No, it's a kinda weird interface the bitset has for iteration, but `next_set` takes `j` by reference and advances it to the next value at-or-after `j` with a set bit, then returns true (or false if it's out of bits).

Reply:
thanks for explanation, but it does not seem intuitive at all...